src/other/Comms.js

```diff
@@ -188,53 +192,93 @@
+                        .url(url)
+                    ;
+                    url = newUrl + context._proxyMappings[key];
+                    request.IP = espUrl._hostname;
+                    request.PORT = espUrl._port;
+                    if (newUrlParts.length > 0) {
+                        request.PATH = newUrlParts[1];
+                    }
+                    break;
                 }
-                break;
             }
-        }
 
-        var respondedTimeout = 60000;
-        var respondedTick = 5000;
-        var callbackName = "jsonp_callback_" + Math.round(Math.random() * 999999);
-        window[callbackName] = function (response) {
-            respondedTimeout = 0;
-            doCallback(response);
-        };
-        var script = document.createElement("script");
-        script.src = url + (url.indexOf("?") >= 0 ? "&" : "?") + "jsonp=" + callbackName + "&" + serialize(request);
-        document.body.appendChild(script);
-        var progress = setInterval(function () {
-            if (respondedTimeout <= 0) {
-                clearInterval(progress);
-            } else {
-                respondedTimeout -= respondedTick;
+            var respondedTimeout = context.timeout() * 1000;
+            var respondedTick = 5000;
+            var callbackName = "jsonp_callback_" + Math.round(Math.random() * 999999);
+            window[callbackName] = function (response) {
+                respondedTimeout = 0;
+                doCallback(response);
+                resolve(response);
+            };
+            var script = document.createElement("script");
+            script.src = url + (url.indexOf("?") >= 0 ? "&" : "?") + "jsonp=" + callbackName + "&" + serialize(request);
+            document.body.appendChild(script);
+            var progress = setInterval(function () {
                 if (respondedTimeout <= 0) {
                     clearInterval(progress);
-                    console.log("Request timeout:  " + script.src);
-                    doCallback();
                 } else {
-                    console.log("Request pending (" + respondedTimeout / 1000 + " sec):  " + script.src);
+                    respondedTimeout -= respondedTick;
+                    if (respondedTimeout <= 0) {
+                        clearInterval(progress);
+                        console.log("Request timeout:  " + script.src);
+                        doCallback();
+                        reject(Error("Request timeout:  " + script.src));
+                    } else {
+                        console.log("Request pending (" + respondedTimeout / 1000 + " sec):  " + script.src);
+                    }
+                }
+            }, respondedTick);
+
+            function doCallback(response) {
+                delete window[callbackName];
+                document.body.removeChild(script);
+                if (callback) {
+                    console.log("Deprecated:  callback, use promise (Comms.prototype.jsonp)");
```

Comment:
can you explain this? not sure why it would this if using promise? and how do we tie into the promise from the outside